### PR TITLE
UX: warn about units being ignored when inserting Quantity values into non Quantity columns in a QTable

### LINKED ISSUE #15964: Empty QTable strips units from newly added row
### Description

When adding quantities to an empty previously created `QTable` using `add_row`, the units are silently stripped and values are added to the table as `float` objects instead of `Quantity` objects.

### Expected behavior

`units` should be maintained when row is added to empty `QTable`, or an error/warning should be raised.

### How to Reproduce

```>>> from astropy.table import QTable
>>> from astropy import unit as u

>>> qtab = QTable(names=['key_1', 'key_2', 'key_3'])
>>> print(qtab)
key_1 key_2 key_3
----- ----- -----

>>> qtab.add_row([1*u.m, 2*u.m, 3*u.m])
>>> qtab.add_row([4*u.m, 5*u.m, 6*u.m])
>>> print(qtab)
key_1 key_2 key_3
----- ----- -----
  1.0   2.0   3.0
  4.0   5.0   6.0
```

### Versions

_No response_

### PR DESCRIPTION
**Pull request by neutrinoceros**
### Description

Fixes #15964 (second try)
See discussion over at #15971

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

File: docs/changes/16004.other.rst
Comment (pllim) [inline docs/changes/16004.other.rst:1]: Not technically API change, but we do lump behavior change under API... But I'll let the table maintainers decide whether this is okay or should be renamed to `api`.

**Review by mhvk — COMMENTED** [review]
Hmm, wouldn't it make more sense to have the warning in `Column.insert`?

**Comment by neutrinoceros** [comment]
Actually, I'm taking it back. Putting the warning in `Column` would only allow the warning to trigger for the first concerned column (assuming default interpreter behaviour where warnings are emitted at most once per callsite).

We could also have 2 distinct warnings show up at once. Let me try that.

**Comment by neutrinoceros** [comment]
I tried the following
```patch
diff --git a/astropy/table/column.py b/astropy/table/column.py
index 3e53240697..202ad3c371 100644
--- a/astropy/table/column.py
+++ b/astropy/table/column.py
@@ -1407,6 +1407,13 @@ class Column(BaseColumn):
             A copy of column with ``values`` and ``mask`` inserted.  Note that the
             insertion does not occur in-place: a new column is returned.
         """
+        if any(isinstance(v, Quantity) for v in values):
+            warnings.warn(
+                "Units from Quantity values will be ignored.",
+                category=UserWarning,
+                stacklevel=2,
+            )
+
         if self.dtype.kind == "O":
             # Even if values is array-like (e.g. [1,2,3]), insert as a single
             # object.  Numpy.insert instead inserts each element in an array-like
```

only to realise that this breaks 157 tests in `table/tests` alone (by virtue of warnings being treated as errors in tests). I'm not sure that level of noisiness is warranted here.

**Comment by mhvk** [comment]
> this breaks 157 tests in `table/tests` alone

Ouch!

**Comment by taldcroft** [comment]
Catching up here...

> wouldn't it make more sense to have the warning in Column.insert

Agreed.

On the issue of warnings, it looks like we should set this one to issue the warning every time instead of just the first time.

About the stack level, I'm not sure it is feasible to always figure out how to get up to the original calling code. There are libraries that will use insert in various ways that we can't predict. Just doing the simple thing of going up one level might be good enough?

**Comment by neutrinoceros** [comment]
> On the issue of warnings, it looks like we should set this one to issue the warning every time instead of just the first time.

As I started writing my response, I initially thought this wasn't possible, so I researched it and found I was wrong !
```python
for _ in range(2):
    with warnings.catch_warnings():
        warnings.simplefilter("always")
        warnings.warn("this goes off every time", stacklevel=2)
```

> Just doing the simple thing of going up one level might be good enough?

This might indeed be an instance of "practicality beats purity". However, even if there's a consensus that this route should be preferred (it seems like there is), we need to address the elephant in the room that 157 tests are known to break if we introduce the warning in this preferred place. So it's all easily browsable to everyone in the discussion, let me push that change now.

**Comment by neutrinoceros** [comment]
Actually I'm able to solve 148/157 issues by replacing the `try/except` block with a more fine-tuned context manager.
Let's see if I can also resolve the remaining 9.

**Comment by neutrinoceros** [comment]
okay, the solution is suddenly much less localised but it's working. The only remaining failure is in the test I'm adding, because I'm only catching one of the two warnings that are currently added. I'm hesitant to completely remove my first warning as it's much more informative, but could be persuaded. Let me know what you think !

File: astropy/table/column.py [hunk 1]
```diff
@@ -1407,6 +1407,30 @@ def insert(self, obj, values, axis=0):
             A copy of column with ``values`` and ``mask`` inserted.  Note that the
             insertion does not occur in-place: a new column is returned.
         """
+
+        def warn_quantities(values):
```
Comment (neutrinoceros) [inline astropy/table/column.py:1411]: note: I've put this in a closure mostly to avoid polluting the scope of the parent function, but this could also be achieved using `del`. Anyhow, this is an implementation detail, we can change it.

**Comment by mhvk** [comment]
Now the bigger comment - I actually wonder why things even work as is, because the actual insertion is done as https://github.com/astropy/astropy/blob/bdfa3eac34a2651664c9735a0a9d7f03c7c7fed5/astropy/table/column.py#L1417-L1418

and `np.insert` is overwritten by `u.Quantity` to check units, https://github.com/astropy/astropy/blob/bdfa3eac34a2651664c9735a0a9d7f03c7c7fed5/astropy/units/quantity_helper/function_helpers.py#L518-L526

This works for arrays,
```
np.insert(a, 1, q)
UnitConversionError: 'm' (length) and '' (dimensionless) are not convertible
```
but somehow not for `Column`,
```
c = Column(np.arange(10.))
np.insert(c, 1, q)
array([0., 1., 1., 2., 3., 4., 5., 6., 7., 8., 9.])
```

I actually do not quite understand why it doesn't work, but do feel this is the correct approach: let `Quantity` tell that it cannot be inserted (if the units do not match).

...

Hmm, `__array_function__` for an `ndarray` lets `Quantity` do its work (because it is an `ndarray` subclass), so that leads to the error, but for `Column`, which is also an `ndarray` subclass, first `Column.__array_function__` is tried. Since that one actually works, the result is bad.  An addition to `BaseColumn` of
```
    def __array_function__(self, function, types, args, kwargs):
        if function is np.insert and Quantity in types:
            return NotImplemented
        return super().__array_function__(function, types, args, kwargs)
```
does indeed cause the error to get raised.

But this is more of an hack than we really want, at least for this PR!

**Comment by mhvk** [comment]
Hmm, running out of time, but this diff does fix things and gives no errors
```
--- a/astropy/table/column.py
+++ b/astropy/table/column.py
@@ -159,7 +159,9 @@ def _expand_string_array_for_values(arr, values):
             arr_dtype = arr.dtype.byteorder + arr.dtype.kind + str(values_str_len)
             arr = arr.astype(arr_dtype)
 
-    return arr
+    return (
+        arr.view(np.ndarray) if isinstance(arr, Column) else arr.view(np.ma.MaskedArray)
+    )
 
 ```
It might be more elegant to do 
```
self_for_insert = _expand_string_array_for_values(self, values).view(np.array)
```
in `Column.insert` and
```
self_ma = _expand_string_array_for_values(self_ma, values).view(np.ma.MaskedArray)
```
in `MaskedColumn.insert`

**Comment by mhvk** [comment]
p.s. Sorry, one last thing: all the above suggestions of `Quantity` taking the lead here are of course different in that inserting a `Quantity` with the wrong unit will actually lead to an error, not just a warning. And `Quantities` will transform themselves to the column unit as needed.

**Comment by mhvk** [comment]
I thought a bit more about it, to see where consistency would lead us. Basically, I think insertion should behave the same as if one inserted a new item, of whatever type the `Column` has, and then set it. As is, it is actually bit of a mess:
```
c = Column(np.arange(10.))
c[1] = 1*u.cm
TypeError: only dimensionless scalar quantities can be converted to Python scalars
```
But
```
c[1:2] = 1*u.cm
```
just works. (This also means that `c[1]=1*u.percent` and `c[1:2] = 1.*u.percent` both work but with different results...). Note that this behaviour is shared with `ndarray`, so in that sense no surprise there.

Right now, given how inconsistent the present behaviour is, my tendency would be to go back to raising a warning inside `Table`, perhaps even just for the `N==0` case, and then perhaps have a larger issue where we consider how `Column` should interact with `Quantity`.

File: astropy/table/column.py [hunk 2]
```diff
@@ -1407,6 +1407,30 @@ def insert(self, obj, values, axis=0):
...
+            isquantity = isinstance(values, Quantity)
+            if not isquantity:
```
[the remark this answers is not in the document]
Comment (neutrinoceros) [reply astropy/table/column.py:1413]: I like how your alternative code isn't specific to `Quantity` ! I'll push this in a couple minutes if it survives local testing !

File: astropy/table/table.py [hunk 3]
```diff
@@ -3274,10 +3275,21 @@ def insert_row(self, index, vals=None, mask=None):
         else:
             raise TypeError("Vals must be an iterable or mapping or None")
 
+        @contextmanager
+        def reraise_true_exceptions_as_ValueError(name: str):
+            with warnings.catch_warnings():
```
[the remark this answers is not in the document]
Comment (neutrinoceros) [reply astropy/table/table.py:3280]: You're right that warnings are not popping up at all now, when they should still cause tests to fail, so my approach isn't even working (yet). However it is not clear to me that ignoring the warning in tests is worth it, and if I need to change 157 tests to workaround it, it feels like a good indication that the new warning will affect many users and might cause a lot of pain downstream. I'll give myself about 30min to study exactly how large of an impact this would do in the test suite and report here.

**Comment by neutrinoceros** [comment]
I opened https://github.com/astropy/astropy/issues/16037 based on your observations.

**Comment by taldcroft** [comment]
@neutrinoceros - I'm a little confused now. I checked out this PR and then reverted `table.py` and `test_table.py` to a2ee8ebd1c, which is the base of this branch. My goal was to look at the famous 157 test errors. But I saw zero related to this PR, only a bunch from a numpy 2.0 deprecation on `product`.

I have always not understood the testing errors because we should not be inserting `Quantity` into `Column`, at least not very often. So ??? I'm sure I'm missing something obvious.

**Comment by neutrinoceros** [comment]
@taldcroft sorry this has become so hard to follow ! What actually happened is that my *condition* for emitting a warning wasn't very robust and would choke on otherwise valid inputs (like 0-d ndarrays). Because this happened within a large `try/except Exception` block, the errors would be reraised as `ValueError` with a generic message, which is also what *would* have happened if the warning was actually raised in a bunch of tests where it wasn't caught. So the confusion is really on me.
I resolved this problem in my most recent commits, so forget about those 157 failures; you're right, we don't insert `Quantities` in `Column` that often 😅 

To further summarise the current state of this branch:
- the remaining failure is in the new test, where I'm capturing only one warning, but I've added two, so one of them gets to fail the test, it could easily be fixed by adjusting the test but ...
- the final diff turned out to be so much bigger than anticipated that we might just prefer the alternative approach in #16038, so I'm not fixing this PR for now because from where I stand it looks likely that it'll just go to the trash :)

**Comment by neutrinoceros** [comment]
Closing this now that #16038 got accepted
